### PR TITLE
[RFC] add journal bindings around sd-journal

### PR DESCRIPTION
**Pull request by vcaputo**
Work in progress. Based on Dan Mace's initial prototype announced here:
http://lists.freedesktop.org/archives/systemd-devel/2014-June/019840.html

Throwing up to start some discussion...

### CODE REVIEW

**Comment by vcaputo** [comment]
Turned this into a PR in response to https://github.com/coreos/go-systemd/pull/107#discussion_r39799486

File: journal/journal.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	}
+
+	return uint64(usec), nil
+}
+
+func (j *Journal) Print(priority int, message string) error {
+	m := C.CString(message)
+	defer C.free(unsafe.Pointer(m))
+
+	j.mu.Lock()
```
Comment (jonboulle) [inline journal/journal.go]: Ehhhhh. I think we should follow Zbigniew's comment @ http://lists.freedesktop.org/archives/systemd-devel/2014-June/020067.html and split out write/read, there's really no need for Print to have access to the journal object here at all (and hence the lock is pointless)

Comment (vcaputo) [inline journal/journal.go]: hm yeah you're right, the lock isn't needed here.

File: journal/read.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+		c, err := r.Read(msg)
+
+		if err != nil && err != io.EOF {
+			return err
+		}
+
+		select {
+		case <-until:
+			fmt.Println("Timed out flushing new entries")
```
Comment (jonboulle) [inline journal/read.go]: maybe return errors.New("timed out flushing new entries") ?

File: journal/read.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+					break poll
+				default:
+					events <- r.journal.Wait(time.Duration(1) * time.Second)
+				}
+			}
+		}()
+
+		select {
+		case <-until:
+			fmt.Println("Timed out waiting for events")
```
Comment (jonboulle) [inline journal/read.go]: as above

File: journal/read.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		}
+
+		// We're at the tail, so wait for new events or time out.
+		// Holds journal events to process. Tightly bounded for now unless there's a
+		// reason to unblock the journal watch routine more quickly.
+		events := make(chan int, 1)
+		pollDone := make(chan bool, 1)
+		go func() {
+		poll:
```
Comment (jonboulle) [inline journal/read.go]: omit this and return on line 144 imo

**Comment by jonboulle** [comment]
@ironcladlou are you happy for us to adapt your code and merge here? we prefer all code is under Apache 2.0 - in this case we'd assign copyright to CoreOS and Red Hat, if that suits you.

**Comment by ironcladlou** [comment]
@jonboulle 
> are you happy for us to adapt your code and merge here? we prefer all code is under Apache 2.0 - in this case we'd assign copyright to CoreOS and Red Hat, if that suits you.

I'll follow up on this internally at Red Hat tomorrow and get back to you ASAP. Thanks!

**Comment by ironcladlou** [comment]
@jonboulle 

Sorry for the delayed response. I have no issue with the journald work being licensed under ASL2 with an assignment to CoreOS/Red Hat. Thanks!

**Comment by jonboulle** [comment]
@ironcladlou thanks so much!

@vcaputo would you mind updating w/copyright header now? also I suggest we just disable the test on travis as it will be impossible to get it working there.

**Comment by jonboulle** [comment]
@vcaputo ping?

**Comment by vcaputo** [comment]
gdmfsob

**Comment by jonboulle** [comment]
someone pass the HO2CCH(NH2)CH2SH

**Comment by vcaputo** [comment]
@jonboulle travis is happy, do you want me to merge this or is there more discussion needed?

**Comment by jonboulle** [comment]
LGTM, let's do it.

**Pull request merged**